[#161996] Stepped billing Documentation (#3825)
tablexi/nucore-open: doc/product_pricing.md

# Product Pricing
## All products
### Price rules

> Products have different price groups, that can either be internal or external. Based on the user's price group and the rate set for it in the product price rules, the final charge will be different.

Reviewer: ```suggestion
Products are assigned to price rules which allow setting rates associated with different price groups, which can either be internal or external. Based on the user's price group and the rate set for it in the product price rules, the final charge will be different.
```

#### Internal

> For internal price groups there is a base rate. There can be adjustments set for other internal price groups.

Reviewer: ```suggestion
For internal price groups there is a base rate. An adjustment amount can be configured for other internal price groups. This amount will be subtracted from the base rate."
```

#### External

> External price groups have a different rate and adjustments cannot be set.

Reviewer: ```suggestion
External price groups have their own rates and adjustments cannot be set.
```

### Charge for

#### Reservation

Users are charged for the length of their reservation.

_Example_

Reservation 1-2pm

Result: User will always be charged 60 minutes, no matter the actual usage.

#### Usage

Users are charged for the length of time they actually used the instrument.

This should be used in conjunction with either "Timer without relay" or "Timer with relay"
control mechanism set on the Instrument detail edit screen.

_Examples_

Reservation 1-2pm

Usage: 1:15-1:45pm
Result: User is charged for 30 minutes

Usage: 1:00-2:15pm
Result: User is charged for 75 minutes

Usage: 1:15-2:15pm
Result: User is charged for 60 minutes

#### Overage

Users are charged for the length of their reservation plus any additional time spent
using the instrument after the end of the reservation.

This should be used in conjunction with either "Timer without relay" or "Timer with relay"
control mechanism set on the Instrument detail edit screen.

_Examples_

Reservation 1-2pm

Usage: 1:15-1:45pm
Result: User is charged for 60 minutes (the reservation length)

Usage: 1:00-2:15pm
Result: User is charged for 75 minutes (reservation plus 15 minutes of overage)

Usage: 1:15-2:15pm

> Result: User is charged for 75 minutes (reservation plus 15 minutes of overage)

Reviewer: Is this correct?  I would think this one would be 60 min

Author: This is part of what I just copied from the previous file, but I think you're right.

### Pricing mode
#### Schedule rule/price group based
Schedule rules state what days and times an instrument is available for reservation.

> When using this Priicng mode, the user will pay an hourly rate defined by the price group they belong to.

Reviewer: ```suggestion
When using this Pricng mode, the user will pay an hourly rate defined by the Price Group they belong to.  Admins can set a percentage discount for each global Price Group during the times defined by each schedule rule.
```

#### Duration based

> In this mode, there are up to 4 rates for each price policy. These rates apply to a given step, which is defined by setting a Rate start (hr). Price policies can either have no steps defined (just the usage rate) or all of them. It's invalid to set two or three steps.

Reviewer: ```suggestion
In this mode, sometimes referred to as "Stepped Billing" mode, there are up to 4 rates for each price policy. These rates apply to a given step, which is defined by setting a Rate start (hr). Price policies can either have no steps defined (just the usage rate) or all of them. It's invalid to set rates for one or two steps.
```

Author: I found this tricky. We say there are 4 steps (usage rate/subsidy + 3 duration rates) but then we want to say you either set all or none of them. I wonder what's the best way to be clear about this.

Maybe something like _It's invalid to set rates for one or two *additional* steps._?

> The first step will always start at 0 hours and its rate will be price policy's `usage_rate`. Note: This step is not created in the DB as a Rate Start, is inferred.

Reviewer: ```suggestion
The first step, also called the "Initial Rate", will always start at 0 hours and its rate will be price policy's `usage_rate`. Note: This step is not created in the DB as a Rate Start, it is inferred from the price policy record.
```

> Step final price for internal groups is calculated as `Base step rate - Group step adjustment`.

Reviewer: ```suggestion
Step hourly rate for non-base internal groups is calculated as `Base step rate - step adjustment`.  See examples below for more info.
```

|-----------------|--------------|---------------------|--------------------|--------------------|
| Base Rate       | 50	         | Rate per hr: 45     | Rate per hr: 40    |	Rate per hr: 39    |
| Other internal  | 5     	     | Adjustment:   8	   | Adjustment:  10	  | Adjustment:  12    |
| External        | 65           | Rate per hr: 62	   | Rate per hr: 60    | Rate per hr: 60    |
| Other External  | 70           | Rate per hr: 64	   | Rate per hr: 62	  | Rate per hr: 58    |

In this example, users that are in the Base Rate Price Group will pay $50/hr for the first 2 hours, $45/hr for the next 3 hours (2-5), $40/hr for the next 2 hours (5-7), and $39/hr for all the rest (7+ hours).

Reservation for 10 hours:
  2 * 50
+ 3 * 45
+ 2 * 40
+ 3 * 39
= 100 + 135 + 80 + 117 = 432

On the other hand, users in Other Internal Price Group will have an adjustment of $5/hr for the first 2 hours, $8/hr for the next 3 hours (2-5), $10/hr for the next 2 hours (5-7), and $12/hr for the rest (7+ hours).

Reservation for 10 hours:
  2 * (50 - 5)
+ 3 * (45 - 8)
+ 2 * (40 - 10)
+ 3 * (39 - 12)
= 90 + 74 + 60 + 81 = 305

### Cancel/reservation charges

> Reservations have a Reservation Cost, that is only invoked if user fail to cancel it within a set time window. For some instruments, there may not be a window, so there are charges when cancelling, regardless of the closeness to the reservation start time.

Reviewer: Is this accurate?
```suggestion
Admins can set a Reservation Cost in the price rule, which is only invoked if user fail to cancel their reservation within a set time window. If not window is set, cancelation charges are applied regardless of the closeness to the reservation start time.
```
